raft/src/main/java/org/apache/kafka/raft/LeaderState.java

```diff
@@ -109,50 +114,86 @@ private OptionalLong quorumMajorityFetchTimestamp() {
      *         not received fetch from the majority yet
      */
     public OptionalLong updateFetchTimestamp(int nodeId, long timestamp) {
-        ReplicaState state = ensureValidVoter(nodeId);
+        NodeState state = ensureValidVoter(nodeId);
         // To be resilient to system time shifts we do not strictly require the timestamp be monotonically increasing
         state.lastFetchTimestamp = OptionalLong.of(Math.max(state.lastFetchTimestamp.orElse(-1L), timestamp));
         return quorumMajorityFetchTimestamp();
     }
 
     public List<Integer> nonLeaderVotersByDescendingFetchOffset() {
         return followersByDescendingFetchOffset().stream()
-                   .filter(state -> state.nodeId != localId)
-                   .map(state -> state.nodeId).collect(Collectors.toList());
+                   .filter(state -> state.replicaId() != localId)
+                   .map(ReplicaState::replicaId).collect(Collectors.toList());
     }
 
     private List<ReplicaState> followersByDescendingFetchOffset() {
-        List<ReplicaState> followersByDescendingFetchOffset = new ArrayList<>(this.voterReplicaStates.values());
-        Collections.sort(followersByDescendingFetchOffset);
-        return followersByDescendingFetchOffset;
+        return new ArrayList<>(this.voterReplicaStates.values())
+            .stream().sorted().map(voterState -> new ReplicaState()
```

Comment:
Not sure I understand why it's necessary to convert here.

Reply:
Removed